# NLP Lab : Modèles de langue

Dans ce tp, nous allons constuire les briques principales du modèle GPT2 et entrainer un petit modèle sur des poèmes de Victor Hugo. 

Les questions sont posées dans ce notebook, puis pour executer l'entrainement, il faudra modifier le ficher `gpt_single_head.py` aussi disponible dans le reposository git.


## Données

Les données d'entrainement sont un recueil de poèmes de Victor Hugo issu du site [gutenberg.org](https://www.gutenberg.org/). Elles sont disponibles dans le répertoire `data`.

Afin de réduire la complexité du modèle, nous allons modéliser le texte au niveau caractère. Les modèles de language modélisent généralement des séquences de sous-mots en utilisant des [tokenizers](https://huggingface.co/docs/transformers/tokenizer_summary) (BPE, SentencePiece, WordPiece)

Questions :
>* En utilisant [collections.Counter](https://docs.python.org/3/library/collections.html#collections.Counter), afficher le nombre de caractères différents dans le texte et la fréquence de chaque caractère.

0.009989 M parameters
step 0: train loss 4.6896, val loss 4.6661
step 500: train loss 0.5564, val loss 0.5934
step 1000: train loss 0.4698, val loss 0.4759
step 1500: train loss 0.4441, val loss 0.4480
step 2000: train loss 0.4219, val loss 0.4275
step 2500: train loss 0.4204, val loss 0.4200
step 3000: train loss 0.4118, val loss 0.4103
step 3500: train loss 0.4056, val loss 0.4098
step 4000: train loss 0.3988, val loss 0.3949
step 4500: train loss 0.4033, val loss 0.3929
step 4999: train loss 0.3988, val loss 0.3885

LLLLLLLLLLLLLL'LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLoLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLaLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLIILLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL

<module 'gpt_single_head' from '/home/onyxia/work/nlp-lab_language-models/gpt_single_head.py'>

In [4]:
import collections

with open('data/hugo_contemplations.txt', 'r', encoding='utf-8') as f:
    text = f.read()

print(f'Number of characters in the file: {len(text)}')
##  YOUR CODE HERE

counter = collections.Counter(text)

###

print (f'Number of character in counter: {sum(counter.values())}')
print (f'{len(counter)} different characters')
print (counter)


Number of characters in the file: 285222
Number of character in counter: 285222
101 different characters
Counter({' ': 49127, 'e': 30253, 's': 17987, 'u': 14254, 'r': 14223, 't': 14071, 'a': 14048, 'n': 13725, 'i': 12828, 'o': 12653, 'l': 11638, '\n': 8102, 'm': 6495, 'd': 6375, ',': 6077, 'c': 5074, 'p': 4206, "'": 3820, 'v': 3492, 'é': 2943, 'b': 2783, 'f': 2772, 'h': 2221, 'q': 1956, 'g': 1790, '.': 1420, 'x': 1154, 'L': 1147, '!': 1121, 'E': 1074, ';': 1043, '-': 1020, 'j': 890, 'D': 764, 'è': 725, 'à': 706, 'y': 660, 'I': 627, 'ê': 605, 'C': 593, 'S': 545, 'A': 530, 'Q': 503, 'z': 482, 'J': 471, 'O': 450, 'T': 441, 'P': 435, '?': 388, 'V': 383, 'â': 381, 'N': 362, 'M': 344, 'ù': 298, ':': 294, 'R': 240, 'î': 214, 'U': 208, 'ô': 159, 'X': 150, '1': 146, 'H': 116, 'F': 114, '5': 111, '8': 93, 'B': 78, '«': 74, 'É': 70, '»': 69, 'G': 67, '4': 64, 'û': 62, '3': 47, 'ç': 34, 'À': 33, 'ë': 32, 'ï': 31, '2': 30, '·': 26, 'Ê': 24, '6': 23, '7': 23, 'Ô': 19, '9': 19, 'È': 11, 'k': 10, '0':

In [35]:
# here are all the unique characters that occur in this text
chars = sorted(list(set(text)))
vocab_size = len(chars)
# create a mapping from characters to integers
stoi = { ch:i for i,ch in enumerate(chars) }
itos = { i:ch for i,ch in enumerate(chars) }
encode = lambda s: [stoi[c] for c in s] # encoder: take a string, output a list of integers
decode = lambda l: ''.join([itos[i] for i in l]) # decoder: take a list of integers, output a string

### Encodage / décodage
Afin de transformer le texte en vecteur pour le réseau de neurones, il faut encoder chaque caractère avec un entier. Les fonctions suivantes opérent l'encodage et le décodage des caractères :

In [5]:
# create a mapping from characters to integers
stoi = { ch:i for i,ch in enumerate(counter) }
itos = { i:ch for i,ch in enumerate(counter) }
encode = lambda s: [stoi[c] for c in s] # encoder: transform a string into a list of integers
decode = lambda l: ''.join([itos[i] for i in l]) # decoder: transform a list of integers into a string


# test that your encoder/decoder is coherent
testString = "\nDemain, dès l'aube"
assert decode(encode (testString)) ==  testString

### Découpage Train/Validation

L'objectif étant de prédire des poèmes, il ne faut pas mélanger les lignes aléatoirement. Il faut garder l'ordre des lignes dans le texte et uniquement prendre les premiers 90% pour entrainer et les 10% restants pour contrôler l'apprentissage. 

Questions :
> * Découper en `train_data` (90%) et `val_data` (10%) en utilisant du slicing sur data.

In [8]:
#!pip install torch

In [9]:
import torch
# Train and validation splits
data = torch.tensor(encode(text), dtype=torch.long)
## YOUR CODE HERE
# first 90% will be train, rest val
# Calcul de l'indice de séparation
split_idx = int(0.9 * len(data))

# Division desz données en ensembles d'entraînement et de validation
train_data = data[:split_idx]
val_data = data[split_idx:]
###

### Contexte

Le modèle de langue possède comme paramètre la taille maximale du contexte à considérer pour faire la prédiction du prochain caractère. Ce contexte est appelé `block_size`. Les données d'apprentissage sont donc des séquences de caractères consécutifs, issues de l'ensemble d'entraînenement tirées aléatoirement et de longueur `block_size`.

Si le caractère de début de la séquence est `i`, la séquence de contexte est donc :
``` x = data[i:i+block_size]```
et la valeur à prédire à chaque position dans le contexte est le caractère suivant :
```y = [data[i+1:i+block_size+1]```.



In [10]:
block_size = 8

i  = torch.randint(len(data) - block_size, (1,))
print (i)
x = train_data[i:i+block_size]
y = train_data[i+1:i+1+block_size]

for t in range(block_size):
    context = x[:t+1]
    target = y[t]
    print (f'context is >{decode(context.tolist())}< target is >{decode([target.tolist()])}<')

tensor([175930])
context is >é< target is >j<
context is >éj< target is >à<
context is >éjà< target is > <
context is >éjà < target is >p<
context is >éjà p< target is >r<
context is >éjà pr< target is >i<
context is >éjà pri< target is >s<
context is >éjà pris< target is > <


### Définition des batchs

Les batchs d'entrainement sont constitués de plusieurs séquences de caractères tirées aléatoirement dans `train_data`. Pour choisir aléatoirement une séquence à mettre dans le batch, il faut tirer aléatoirement un point de départ dans `train_data` et extraire les `block_size` caractères suivants. Lors du tirage du point de départ, faire attention à laisser suffisamment de caractères après le point de départ pour avoir une séquence de `block_size` caractères.

Questions :
> * Créer les batchs `x` en tirant `batch_size` séquences de longeur `block_size` à  partir d'un point `i` tiré aléatoirement. Empiler les exemples avec `torch.stack`.
> * Créer les batchs `y` en ajoutant le caractère suivant la séquence `x`. Empiler les exemples avec `torch.stack`.




In [18]:
batch_size = 4
torch.manual_seed(2023)
device = torch.device('cuda' if torch.cuda.is_available() else 'cpu')
# data loading
def get_batch(split):
    # generate a small batch of data of inputs x and targets y
    data = train_data if split == 'train' else val_data
    data_len=len(data)
    ## YOUR CODE HERE
    # select batch_size starting points in the data, store them in a list called starting_points
    starting_points = torch.randint(0, data_len - block_size, (batch_size,))
    
    # x is the sequence of integer starting at each straing point and of length block_size
    # y is the character after each starting position
    x_list = []
    y_list = []
    for start in starting_points:
        x_seq = data[start:start + block_size]
        y_seq = data[start + 1:start + block_size + 1]
        x_list.append(x_seq)
        y_list.append(y_seq)

    ### 
    x = torch.stack(x_list)
    y = torch.stack(y_list)
    # send data and target to device
    x, y = x.to(device), y.to(device)
    return x, y

### Premier modèle : un bigramme 

Le premier modèle que nous allons implémenter est un modèle bigramme. Il prédit le caractère suivant uniquement en fonction du caractère courant. Il est possible de stocker ce modèle dans une simple matrice : pour chaque caractère (en ligne), on stocke la distribution de probabilités sur l'ensemble des caractères suivants (en colonne). On peut donc le stocker dans une simple couche [`Embedding`](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html).

Questions :
> * Dans le constructeur, définir une couche Embedding de taille `vocab_size` par `vocab_size`.
> * Dans forward, appliquer la couche d'embedding au batch de idx (`x`).
> * Dans forward, définir la loss comme la `cross_entropy` entre la prédiction et target (`y`).



In [36]:
import torch.nn as nn

# use a gpu if we have one
device = 'cuda' if torch.cuda.is_available() else 'cpu'

# Simple bigram model
class BigramLanguageModel(nn.Module):

    def __init__(self, vocab_size):
        super().__init__()
        # we use a simple vocab_size times vocab_size tensor to store the probabilities 
        # of each token given a single token as context in nn.Embedding
        # YOUR CODE HERE
        self.token_embedding_table = nn.Embedding(vocab_size, vocab_size)
        ## 
        
    def forward(self, idx, targets=None):

        # idx and targets are both (Batch,Time) tensor of integers
        # YOUR CODE HERE
        logits = self.token_embedding_table(idx)

        ## 
   
        # don't compute loss if we don't have targets
        if targets is None:
            loss = None
        else:
            # change the shape of the logits and target to match what is needed for CrossEntropyLoss
            # https://pytorch.org/docs/stable/generated/torch.nn.CrossEntropyLoss.html
            Batch, Time, Channels = logits.shape
            logits = logits.view(Batch*Time, Channels)
            targets = targets.view(Batch*Time)
            
            # negative log likelihood between prediction and target
            # YOUR CODE HERE
            loss = nn.functional.cross_entropy(logits, targets)

            ## 

        return logits, loss

    def generate(self, idx, max_new_tokens):
        # idx is (B, T) array of indices in the current context
        for _ in range(max_new_tokens):
            # get the predictions
            logits, loss = self(idx)
            # focus only on the last time step
            logits = logits[:, -1, :] # becomes (B, C)
            # apply softmax to get probabilities
            probs = nn.functional.softmax(logits, dim=-1) # (B, C)
            # sample from the distribution
            idx_next = torch.multinomial(probs, num_samples=1) # (B, 1)
            # append sampled index to the running sequence
            idx = torch.cat((idx, idx_next), dim=1) # (B, T+1)
        return idx

model = BigramLanguageModel(vocab_size)
# send the model to device
m = model.to(device)

#### Modèle avant entraînement
Le modèle n'a pas encore été entrainé, il est juste initialisé, mais on peut calculer la loss sur un batch aléatoire. Les poids étant initialisés avec une distribution normale N(0,1) sur chaque dimension, la loss attendue après l'initialisation devrait être proche de `-ln(1/vocab_size)` (l'entropie est maximale).

In [37]:
import math
xb, yb = get_batch('train')
logits, loss = m(xb, yb)
print (logits.shape)
print (f'loss attendue {-math.log(1.0/vocab_size)}')
print (f'loss calculée {loss}')

torch.Size([32, 101])
loss attendue 4.61512051684126
loss calculée 4.87128210067749


Pour utiliser le modèle en prédiction, il faut lui fournir un premier caractère pour amorcer la séquence : c'est le prompt. Dans notre cas, on peut initialiser la génération avec le caractère de retour à la ligne pour débuter une nouvelle phrase.

Questions :
> * Créer un prompt avec un tenseur de taille (1,1) contenant l'entier correspondant au caractère `\n`.
> * Générer une séquence de caractères de taille 100 à partir de ce prompt avec les fonctions `m.generate` et `decode`.
> * Comment est la phrase générée ?

In [40]:
print (encode(['\n']))
## YOUR CODE HERE
#  le prompt initial
newline_char_code = 100  # Code ASCII pour '\n'
prompt = torch.tensor([[newline_char_code]], dtype=torch.long).to(device)
# Génération la séquence de 100 caractères
generated_indices = m.generate(prompt, max_new_tokens=100)

# decode la séquence générée
generated_text = decode(generated_indices[0].tolist())
print(generated_text)

###

[0]
ûS·X:B;u2»lX'G-hR-E»GuSÂFëh tU'fXÊjdb7èyî56ÊëçmLX8yK«07û;ueÊ3;ÀCÀÎBGB»gbÂÆYBcêWrË:9-hNèrEV«»Uu(WZ ,WE


### Entrainement

Pour l'entrainement, nous utilisons un optimiseur [AdamW](https://pytorch.org/docs/stable/generated/torch.optim.AdamW.html) avec un learning rate de 1e-3. Une itération d'apprentissage consiste en 
- générer un batch
- appliquer le réseau de neurones (forward) et calculer la loss (`model(xb, yb)`)
- calculer le gradient (après avoir remis à zero le gradient cumulé) (`loss.backward()`)
- mettre à jour les paramètres (`optimizer.step()`).

In [41]:
max_iters = 100
batch_size = 4
eval_interval = 10
learning_rate = 1e-3
eval_iters = 20

@torch.no_grad() # no gradient is computed here
def estimate_loss():
    """ Estimate the loss on eval_iters batch of train and val sets."""
    out = {}
    model.eval()
    for split in ['train', 'val']:
        losses = torch.zeros(eval_iters)
        for k in range(eval_iters):
            X, Y = get_batch(split)
            logits, loss = model(X, Y)
            losses[k] = loss.item()
        out[split] = losses.mean()
    model.train()
    return out

# re-create the model
model = BigramLanguageModel(vocab_size)
m = model.to(device)

# create a PyTorch optimizer
optimizer = torch.optim.AdamW(model.parameters(), lr=learning_rate)

for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()


step 0: train loss 4.8717, val loss 5.0225
step 10: train loss 4.9009, val loss 4.9663
step 20: train loss 4.9277, val loss 4.9504
step 30: train loss 4.8989, val loss 4.9776
step 40: train loss 4.8667, val loss 4.9980
step 50: train loss 4.9045, val loss 4.9832
step 60: train loss 4.9071, val loss 5.0338
step 70: train loss 4.8695, val loss 5.0400
step 80: train loss 4.8456, val loss 4.8420
step 90: train loss 4.8983, val loss 4.9423


Une fois le réseau entrainé pendant 100 itérations, on peut générer une séquence de caractères.

Questions :
> * Quel est l'effet de l'entraînement ?
> * Augmenter le nombre d'itérations à 1000 puis à 10,000, noter la loss obtenue et la phrase générée. Qu'observez-vous ?

In [43]:
max_iters=1000
for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

step 0: train loss 4.8650, val loss 4.9508
step 10: train loss 4.8160, val loss 4.9477
step 20: train loss 4.7904, val loss 4.8840
step 30: train loss 4.7969, val loss 4.8855
step 40: train loss 4.8067, val loss 4.8680
step 50: train loss 4.8017, val loss 4.8618
step 60: train loss 4.6901, val loss 4.9335
step 70: train loss 4.7938, val loss 4.7730
step 80: train loss 4.7108, val loss 4.8237
step 90: train loss 4.7469, val loss 4.9137
step 100: train loss 4.7328, val loss 4.8738
step 110: train loss 4.7377, val loss 4.8978
step 120: train loss 4.6909, val loss 4.9682
step 130: train loss 4.6662, val loss 4.8040
step 140: train loss 4.6198, val loss 4.7778
step 150: train loss 4.6633, val loss 4.8052
step 160: train loss 4.7182, val loss 4.7064
step 170: train loss 4.7049, val loss 4.8191
step 180: train loss 4.6808, val loss 4.7833
step 190: train loss 4.7259, val loss 4.7871
step 200: train loss 4.6241, val loss 4.6617
step 210: train loss 4.6895, val loss 4.7454
step 220: train loss 

In [44]:
max_iters=10000
for iter in range(max_iters):

    # every once in a while evaluate the loss on train and val sets
    if iter % eval_interval == 0:
        losses = estimate_loss()
        print(f"step {iter}: train loss {losses['train']:.4f}, val loss {losses['val']:.4f}")

    # sample a batch of data
    xb, yb = get_batch('train')

    # evaluate the loss
    logits, loss = m(xb, yb)
    optimizer.zero_grad(set_to_none=True)
    loss.backward()
    optimizer.step()

step 0: train loss 4.0778, val loss 4.2505
step 10: train loss 4.1174, val loss 4.1878
step 20: train loss 4.0888, val loss 4.2132
step 30: train loss 4.0299, val loss 4.1802
step 40: train loss 4.0360, val loss 4.0778
step 50: train loss 4.0644, val loss 4.2081
step 60: train loss 3.9893, val loss 4.1738
step 70: train loss 4.0772, val loss 4.1078
step 80: train loss 3.9899, val loss 4.2483
step 90: train loss 4.0323, val loss 4.1886
step 100: train loss 4.0532, val loss 4.1688
step 110: train loss 4.0167, val loss 4.2263
step 120: train loss 4.0494, val loss 4.0567
step 130: train loss 4.0354, val loss 4.1471
step 140: train loss 4.0321, val loss 4.1793
step 150: train loss 3.9817, val loss 4.1565
step 160: train loss 4.0442, val loss 4.1635
step 170: train loss 3.9799, val loss 4.0894
step 180: train loss 3.9342, val loss 4.0778
step 190: train loss 4.0771, val loss 4.2122
step 200: train loss 3.9332, val loss 4.1919
step 210: train loss 3.9836, val loss 4.1061
step 220: train loss 

In [42]:
idx = torch.ones((1,1), dtype=torch.long)*3
print (decode(m.generate(idx, max_new_tokens=100)[0].tolist()))

'u9Â1ÉhîÉlâX1LÔbM?SôMeb]ÈYgSi3.6PwOp6Ë:'ç1aÉX Ék'KiéWÈS,ëE)Câ;P'Bs·P[(,ôB,âVÆwUP';ÉrSheZ,EÀ0(ZÉPdy_6k


## Single Head Attention

Nous allons maintenant implémenter le mécanisme de base de l'attention. Pour chaque couple de mots de la séquence, ce mécanisme combine Q une *query* (l'information recherchée), K une *key* (l'information obtenue) et calcule  V une *value*, un vecteur de résultat. 

![single head attention](images/single_head_attention.png)

### Masquage
Cependant, comme nous utilisons le modèle pour générer des séquences, on ne doit pas utiliser les caractères situés après le caractère courant, car ce sont justement ces caractères que l'on cherche à prédire lors de l'apprentissage : *le futur n'est pas utilisé pour prédire (le futur).* 

On va donc intégrer une matrice de masquage dans le processus. Cette matrice indique que pour le premier caractère de la séquence, on ne peux utiliser que ce caractère pour prédire (pas de contexte). Pour le second caractère, on peut utiliser le premier caractère et le second. Pour le troisième caractère, on peut utiliser le premier, le second et le troisième et ainsi de suite. Cette matrice est donc une matrice triangulaire inférieure. Cette matrice est normalisée par ligne (les lignes somment à 1).

In [45]:
T = 8

# first version of the contraints with matrix multiplication
# create a lower triangular matrix
weights0 = torch.tril(torch.ones(T,T))
# normalize each row
weights0 = weights0 / weights0.sum(1, keepdim=True) 
print (weights0)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5000, 0.5000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3333, 0.3333, 0.3333, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2000, 0.2000, 0.2000, 0.2000, 0.2000, 0.0000, 0.0000, 0.0000],
        [0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.0000, 0.0000],
        [0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.0000],
        [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]])


La couche [`softmax`](https://pytorch.org/docs/stable/generated/torch.nn.functional.softmax.html) est une autre manière de réaliser la normalisation :

Question :
> * Vérifier qu'on obtient bien la même matrice.

In [46]:
tril = torch.tril(torch.ones(T,T))
weights = torch.tril(torch.ones(T,T))
weights = weights.masked_fill(tril== 0, float('-inf'))
weights = nn.functional.softmax(weights, dim=-1)
print (weights)

tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.5000, 0.5000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.3333, 0.3333, 0.3333, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2500, 0.2500, 0.2500, 0.2500, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.2000, 0.2000, 0.2000, 0.2000, 0.2000, 0.0000, 0.0000, 0.0000],
        [0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.1667, 0.0000, 0.0000],
        [0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.1429, 0.0000],
        [0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250, 0.1250]])


### Implémentation

Nous pouvons maintenant implémenter la couche d'attention :

![attention_formula](images/attention_formula.png)

Questions :

> * Créer les couches key, query et value comme des couches linéaires de dimension `C` x `head_size`.
> * Appliquer les couches à `x`.
> * `weights = query x key` (transposer les deuxième et troisième dimensions de key pour pouvoir faire le produit).
> * Appliquer le facteur de normalisation.
> * Appliquer le masque triangulaire et la softmax à `weights`.
> * Appliquer value à `x`.
> * Le résultat `out` est la multiplication de `weights` par `value(x)`.


In [48]:
head_size = 16
B, T, C = 4, 8, 32
x = torch.randn(B, T, C)
## YOUR CODE HERE
import torch.nn.functional as F
# define the Key layer  
key = torch.nn.Linear(C, head_size) 
# define the Query layer
query = torch.nn.Linear(C, head_size)
# define the Value layer
value =  torch.nn.Linear(C, head_size)
# apply each layer to the input
k =key(x)  # (B, T, head_size)
q = query(x) # (B, T, head_size)
v = value(x) # (B, T, head_size)
# compute the normalize product between Q and K 
weights = torch.bmm(q, k.transpose(1, 2)) # (B, T, head_size) @ (B, 16, head_size) -> (B, T, T)
# apply the mask (lower triangular matrix)
weights = weights.masked_fill(tril== 0, float('-inf'))
# apply the softmax
weights = F.softmax(weights, dim=-1)
###
out  = weights @ value(x) # (B, T, head_size)

# print the result
print("Weights[0]:\n", weights[0])  
print("Output[0]:\n", out[0])

Weights[0]:
 tensor([[1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.7708, 0.2292, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.6764, 0.1056, 0.2180, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.0493, 0.1134, 0.0035, 0.8338, 0.0000, 0.0000, 0.0000, 0.0000],
        [0.0843, 0.0794, 0.4275, 0.2816, 0.1272, 0.0000, 0.0000, 0.0000],
        [0.5064, 0.1615, 0.2878, 0.0021, 0.0289, 0.0133, 0.0000, 0.0000],
        [0.2550, 0.0683, 0.3723, 0.0511, 0.0904, 0.1244, 0.0384, 0.0000],
        [0.2489, 0.0836, 0.0315, 0.0294, 0.1417, 0.3339, 0.0553, 0.0757]],
       grad_fn=<SelectBackward0>)
Output[0]:
 tensor([[-0.2341, -1.0270, -0.2843, -0.3103,  0.5312,  0.1632,  0.5229,  0.5325,
          0.2807,  0.0490,  1.1303,  0.1615, -0.7270,  0.4876,  0.3824, -0.2562],
        [-0.1534, -0.6405, -0.1427, -0.5283,  0.2434,  0.0489,  0.5092,  0.3375,
          0.2537,  0.1122,  0.7364,  0.1209, -0.3379,  0.4202,  0.2837, -0.1775],
        [-0.0461, -0.3

Questions :

> * Copier votre code dans `gpt_single_head.py` : la définition des couches dans le constructeur de la classe `Head` et les calculs dans la fonction `forward`.
> * Faire un entrainement.
> * Quelle loss en train et val obtenez vous ? Le texte vous parait-il meilleur ?


In [55]:
from importlib import reload
import gpt_single_head as G
reload(G)

0.009989 M parameters
step 0: train loss 4.6896, val loss 4.6661
step 500: train loss 0.5564, val loss 0.5934
step 1000: train loss 0.4698, val loss 0.4759
step 1500: train loss 0.4441, val loss 0.4480
step 2000: train loss 0.4219, val loss 0.4275
step 2500: train loss 0.4204, val loss 0.4200
step 3000: train loss 0.4118, val loss 0.4103
step 3500: train loss 0.4056, val loss 0.4098
step 4000: train loss 0.3988, val loss 0.3949
step 4500: train loss 0.4033, val loss 0.3929
step 4999: train loss 0.3988, val loss 0.3885

LLLLLLLLLLLLLL'LLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLoLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLaLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLIILLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLLL

<module 'gpt_single_head' from '/home/onyxia/work/nlp-lab_language-models/gpt_single_head.py'>

## Multi-head attention

La *multi-head attention* est simplement le calcul en parallèle de plusieurs *single head attention*. Chacune des single head attention est concaténée pour créer la sortie de la multi-head attention. Dans la figure issue de l'article original, le nombre de *heads* dans le *multi-head* est `h`. Afin d'opérer des combinaisons pondérées sur la sortie de chacune des single head, une couche de calcul linéaire est ajoutée.

![multi head attention](images/multi_head_attention.png)

Le code ci-dessous crée un module de multi-head attention.
Questions :
> * Dans le constructeur, créer une liste contenant `num_heads` module `Head` en utilisant la fonction [ModuleList](https://pytorch.org/docs/stable/generated/torch.nn.ModuleList.html) de pytorch.
> * Dans la fonction `forward`, appliquer chaque single head à l'input et concaténer le résultat en utilisant la fonction [cat](https://pytorch.org/docs/stable/generated/torch.cat.html) de pytorch.

In [50]:


class Head(nn.Module):
    def __init__(self, head_size):
        super().__init__()
        self.head_size = head_size
        self.key = nn.Linear(head_size, head_size)
        self.query = nn.Linear(head_size, head_size)
        self.value = nn.Linear(head_size, head_size)
        
    def forward(self, x):
        # Applique les transformations de Key, Query et Value
        k = self.key(x)  # (B, T, head_size)
        q = self.query(x)  # (B, T, head_size)
        v = self.value(x)  # (B, T, head_size)
        
        # Calcul du produit de Q et K pour obtenir les poids d'attention
        weights = torch.bmm(q, k.transpose(1, 2))  # (B, T, head_size) @ (B, head_size, T) -> (B, T, T)
        
        # Appliquer softmax sur les poids pour obtenir la distribution d'attention
        weights = torch.nn.functional.softmax(weights, dim=-1)  # (B, T, T)
        
        # Appliquer les poids d'attention sur les valeurs
        out = torch.bmm(weights, v)  # (B, T, T) @ (B, T, head_size) -> (B, T, head_size)
        
        return out







class MultiHeadAttention(nn.Module):
    """ multiple heads of self-attention in parallel """

    def __init__(self, num_heads, head_size):
        super().__init__()
        ## YOUR CODE HERE
        ## list of num_heads modules of type Head
        self.heads = nn.ModuleList([Head(head_size) for _ in range(num_heads)])
        ###
        
        
    def forward(self, x):
        ## YOUR CODE HERE
        ## apply each head in self.heads to x and concat the results 
        out = torch.cat([h(x) for h in self.heads], dim=-1)

        return out


Questions :
> * Copier le fichier `gpt_single_head.py` en `gpt_multi_head.py`.
> * Ajouter le module MultiHeadAttention dans `gpt_multi_head.py`.
> * En tête de fichier, ajouter un paramètre  `n_head = 4`.
> * Dans le module BigramLanguageModel, remplacer le module Head par un module MultiHeadAttention avec les paramètres `num_heads = n_head` et `head_size = n_embd // n_head` pour garder le même nombre de paramètres.
> Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.009893 M parameters
step 4999: train loss 2.1570, val loss 2.1802

## Ajout d'une couche de calcul FeedForward


Après les couches d'attention qui collectent l'information dans la séquence, une couche de calcul est ajoutée pour combiner toutes les informations de la séquence. Cette couche est un simple Multi-Layer-Perceptron avec une couche cachée et une non linéarité de type [RELU](https://pytorch.org/docs/stable/generated/torch.nn.ReLU.html).
![multi feedfoward](images/multi_ff.png)

In [ ]:
class FeedForward(nn.Module):
    """ a simple MLP with RELU """

    def __init__(self, n_embd):
        super().__init__()
        self.net = nn.Sequential(
            nn.Linear(n_embd, n_embd),
            nn.ReLU(),
        )

    def forward(self, x):
        return self.net(x)

Questions :
> * Ajouter le module `FeedForward` dans votre fichier `gpt_multi_head.py`.
> * Ajouter cette couche `FeedForward` après la *multi-head attention*.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.010949 M parameters
step 4999: train loss 2.1290, val loss 2.1216

In [65]:
from importlib import reload
import gpt_multi_head as G
reload(G)

0.022661 M parameters
step 0: train loss 4.6491, val loss 4.6626
step 500: train loss 3.1173, val loss 3.1449
step 1000: train loss 3.0738, val loss 3.0825
step 1500: train loss 2.5790, val loss 2.5967
step 2000: train loss 2.4037, val loss 2.4332
step 2500: train loss 2.3256, val loss 2.3694
step 3000: train loss 2.2803, val loss 2.3080
step 3500: train loss 2.2545, val loss 2.2505
step 4000: train loss 2.2303, val loss 2.2269
step 4500: train loss 2.2108, val loss 2.1992
step 4999: train loss 2.2079, val loss 2.1526

Vis bait laus das que aiseumaique son l'éunsenvite on plége dotmeun ar bonr limievis;
Le sobbhex
Le la l'aversque mi ferfais, tomme L'énège:
Sa mon, ves, la veullas qu'un nautis est toites eut ufnsaalle esdre.
L'hon -De Des que,
Cies bannes amme et pune-rit jyarte en vaveus, celplofunée daz moitte, cafmon.

             ·
Oleu:
Le la es de voite, Lorrile que ple endovê4mes.

NRIÀ--Jau miose Raniéte; pril sonmroit!
Utemie!n hannecssoir, taur, à loioriez qu'In oshaivite,
Ô

<module 'gpt_multi_head' from '/home/onyxia/work/nlp-lab_language-models/gpt_multi_head.py'>

## Empiler les blocs

Le réseau construit jusqu'à présent n'est en fait qu'un bloc du réseau final. Il est maintenant possible d'empiler les blocs de *multi-head attention* pour créer un réseau profond. 

![multi feedfoward](images/multi_bloc.png)


Le code suivant crée un bloc : 

In [ ]:
class Block(nn.Module):
    """ A single bloc of multi-head attention """

    def __init__(self, n_embd, n_head):
        super().__init__()
        head_size = n_embd // n_head
        self.sa = MultiHeadAttention(n_head, head_size)
        self.ffwd = FeedForward(n_embd)

    def forward(self, x):
        x = self.sa(x)
        x = self.ffwd(x)
        return x

Questions :
> * Ajouter le module `Block` dans `gpt_multi_head.py`.
> * Modifier le code de `BigramLanguageModel` pour ajouter 3 `Block(n_embd, n_head=4)` avec un container [Sequential](https://pytorch.org/docs/stable/generated/torch.nn.Sequential.html) à la place de `MultiHeadAttention`et `FeedForward`.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.019205 M parameters
step 4999: train loss 2.2080, val loss 2.2213

## Amélioration de l'entraînement

Si on veut continuer à augmenter la taille du réseau, il est nécessaire d'utiliser des couches permettant d'améliorer l'entraînement et ses capacités de généralisation (réduire le sur-apprentissage). Ces couches sont :
- *skip connections* ou *residual connections*
- les couches de normalisation
- le dropout.


![multi feedfoward](images/multi_skip_norm.png)


Questions :
> * Dans le module Block, ajouter une skip connection en ajoutant l'input dans chaque connexion :
```
        x = x + self.sa(self.ln1(x))
        x = x + self.ffwd(self.ln2(x))
```
> * Dans le module Block, ajouter 2 couches de [LayerNorm](https://pytorch.org/docs/stable/generated/torch.nn.LayerNorm.html) de taille `n_embd` avant la couche de `Multi-Head attention` et avant la `FeedForward`.
> * Après la série de 3 blocs, ajouter une couche de LayerNorm de taille `n_embd`.
> * Définir une variable `dropout = 0.2` en début de fichier et ajouter une couche de [Dropout](https://pytorch.org/docs/stable/generated/torch.nn.Dropout.html) :
>    * Après la couche RELU dans FeedForward
>    * Après la couche de MultiHead dans `MultiHeadAttention`
>    * Après la softmax dans la single head attention `Head`.
> * Relancer l'entrainement et noter le nombre de paramètres et les loss obtenues.

0.019653 M parameters

In [66]:
from importlib import reload
import gpt_multi_head as G
reload(G)

0.023109 M parameters
step 0: train loss 4.7772, val loss 4.7726
step 500: train loss 2.4534, val loss 2.5041
step 1000: train loss 2.2850, val loss 2.2988
step 1500: train loss 2.2197, val loss 2.2232
step 2000: train loss 2.1688, val loss 2.1730
step 2500: train loss 2.1217, val loss 2.1303
step 3000: train loss 2.1042, val loss 2.0819
step 3500: train loss 2.0848, val loss 2.0678
step 4000: train loss 2.0573, val loss 2.0328
step 4500: train loss 2.0536, val loss 2.0334
step 4999: train loss 2.0543, val loss 2.0216

      
                Pe le les mour. la n'us maité,
Pent Muuâter,
Loite ou douts le bris étempress  yeft le cesis preôTez-réte.

Pantroplôt de de danoe aine janfapsuille sais s'hommére que nor-vees,
IVIN
Me das, un l'umes: ceu splos vaans que fugessonnuser pent la taispré, haièvan!
Jolytre au le meré sle «J'ustres
Où plommez
Punt dans à vaâmo.


E plezur la même la ce échasdoù en ble sêve, lant pa parsta chous dre la vez tous de me choix!
Que voiol à bar et ux de cis l

<module 'gpt_multi_head' from '/home/onyxia/work/nlp-lab_language-models/gpt_multi_head.py'>

## Conclusion

Les principaux éléments de GPT2 sont en place, il faut maintenant faire passer le modèle à l'échelle et l'entraîner sur une base de données beaucoup plus grande. Pour comparaison, les paramètres de [GPT2](https://huggingface.co/transformers/v2.11.0/model_doc/gpt2.html) sont : 

* `vocab_size = 50257` : GPT2 modélise des tokens (subwords) alors que nous modélisons des caractères. Pour nous, `vocab_size = 100`.
* `n_positions = 1024` : la taille maximale du contexte. Pour nous, c'est `block_size = 8`.
* `n_embd = 768`:  la dimension des embeddings. Pour nous c'est `n_embd = 32`.
* `n_layer = 12`: le nombre de block. Pour nous c'est 3.
* `n_head = 12`: le nombre de multi-head attention. Pour nous c'est 4.

Au total, GPT2 est composé de 1,500 millions de paramètres et a été entrainé sur 8M de pages web, soit 40 Gb de texte.


```
10.816613 M parameters
step 0: train loss 4.7847, val loss 4.7701
step 4999: train loss 0.2683, val loss 2.1161
time: 31m47.910s   
    
Le pêcheur où l'homme en peu de Carevante
Sa conter des chosses qu'en ses yoitn!

Ils sont là-hauts parler à leurs ténèbres
A ceux qu'on rêve aux oiseaux des cheveux,
Et celus qu'on tourna jamais sous le front;
Ils se disent tu mêle aux univers.
J'ai vu Jean vu France, potte; petits contempler,
Et petié calme au milibre et versait,
M'éblouissant, emportant, écoute, ingorancessible,
On meurt s'efferayait.....--Pas cont âme parle en Apparia!
```